docs/protocols/java_object_graph_spec.md

```diff
@@ -0,0 +1,541 @@
+much more easy to use, but
+also introduce more complexities compared to static serialization frameworks. So the format will be more complex.
+
+Here is the overall format:
+
+```
+| fury header | object ref meta | object class meta | object value data |
+```
+
+The data are serialized using little endian byte order overall. If bytes swap is costly for some object,
+Fury will write the byte order for that object into the data instead of converting it to little endian.
+
+## Fury header
+
+Fury header consists starts one byte:
+
+```
+|     4 bits    | 1 bit | 1 bit | 1 bit  | 1 bit |          optional 4 bytes          |
++---------------+-------+-------+--------+-------+------------------------------------+
+| reserved bits |  oob  | xlang | endian | null  | unsigned int for meta start offset |
+```
+
+- null flag: 1 when object is null, 0 otherwise. If an object is null, other bits won't be set.
+- endian flag: 1 when the system uses little endian, 0 otherwise.
+- xlang flag: 1 when serialization uses xlang format, 0 when serialization uses Fury java format.
+- oob flag: 1 when passed `BufferCallback` is not null, 0 otherwise.
+
+If meta share mode is enabled, an uncompressed unsigned int is appended to indicate the start offset of metadata.
+
+## Reference Meta
+
+Reference tracking handles whether the object is null, and whether to track reference for the object by writing
+corresponding flags and maintaining internal state.
+
+Reference flags:
+
+| Flag                | Byte Value | Description                                                                                                                              |
+|---------------------|------------|------------------------------------------------------------------------------------------------------------------------------------------|
+| NULL FLAG           | `-3`       | This flag indicates the object is a not-null value. We don't use another byte to indicate REF, so that we can save one byte.             |
+| REF FLAG            | `-2`       | This flag indicates the object is already serialized previously, and fury will write a ref id with unsigned varint format instead of serialize it again |
+| NOT_NULL VALUE FLAG | `-1`       | This flag indicates the object is a non-null value and fury doesn't track ref for this type of object.                                   |
+| REF VALUE FLAG      | `0`        | This flag indicates the object is a referencable and first write.                                                                        |
+
+When reference tracking is disabled globally or only for some type, or for some type under some context such as some
+field of a class, only `NULL FLAG` and ` NOT_NULL VALUE FLAG` will be used.
+
+## Class Meta
+
+Fury support register class by an optional id, the registration can be used to security check, and identify the class.
+If the class is registered, the class will have a user-provided or an auto-growing unsigned int `class_id`.
```

Comment:
```suggestion
Fury supports to register class by an optional id, the registration can be used for security check and class identification.
If a class is registered, it will have a user-provided or an auto-growing unsigned int i.e. `class_id`.
```